### ISSUE TITLE
Add a "type" field to changefeed entries

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/rethinkdb#5188]
Every program I've written using changefeeds has had boilerplate akin to the following in it:

```js
if (change.new_val != null && change.old_val == null ) {
  change.type = 'added'
} else if ( change.new_val == null && change.old_val != null ) {
  change.type = 'removed'
} else if (change.new_val != null && change.old_val != null ) {
  change.type = 'changed'
// And the new guys:
} else if (change.new_val != undefined && change.old_val == undefined) {
  change.type = 'initial'
} else if (change.new_val == undefined && change.old_val != undefined) {
  change.type = 'uninitial'
}
```

This is pretty mind-numbing, and I would guess almost always necessary ( @segphault and @dalanmiller have written way more apps than me and could comment). In addition, it's also error prone. (There are two errors I made in the example above, can you spot them?) The drivers can provide convenience wrappers, but I think we should just make the changefeeds themselves more user friendly.

I propose we add a `type` field to changefeed entries, and just put the type in as a string. This would also be a natural place to put `type: "error"` (see #5041).

Cons:
1. slightly more cluttered
2. slightly more bandwidth
3. *sometimes* you can just delete old_val, add the new_val and you don't need to exactly classify the type. Even in this case, you can just ignore the type field.

Pros:
1. Obviously what everyone really wants
2. Come on

(I would also extend this behavior to the results from `return_changes` but I would guess the code is shared and so this would just happen as a side-effect)

**Comment by danielmewes** [comment]
This is almost the same as https://github.com/rethinkdb/rethinkdb/issues/4194, except that the field name and string values are different:

> The proposal is to encode this information explicitly by adding an additional field. For example
> `action: 'delete'`
> `action: 'insert'`
> `action: 'replace'`
> `action: 'initial'`

I'm going to close the other one, because this one has the nicer write up.

Another "Pro" from the other issue:
The lack of a `type`/`action` field sometimes actually makes it impossible to know what happened.
a) Setting the value of a changefeed item to the value `null` and
b) deleting a changefeed value
are currently undistinguishable.

**Comment by deontologician** [comment]
Would this be a lot extra to implement on top of #5041? I.e. could we do it in 2.3?

**Comment by danielmewes** [comment]
@deontologician Did you mean a different issue? I'm not sure how this is related to #5041.

Would this help with Fusion? I'm not opposed to bumping this into 2.3. Just want to understand how urgently we need this.

**Comment by deontologician** [comment]
Oh, sorry, #5041 has the concept of "error" changes, which I suggest become `{type: "error"}`.

I don't think this is "urgent" since there's a workaround everyone is already doing (barring the edge case correctness issues). But it would be a huge interface improvement for (what I would guess) is not a lot of effort

**Comment by danielmewes** [comment]
@deontologician That's true. Let me sneak this in.

**Comment by encryptio** [comment]
I like this, for readability/maintainability's sake.

There's not a lot of difference, but I prefer @deontologician's `"type"` values in the OP (`"added"` etc), because they more clearly imply that this is an event stream, as opposed to the `"action"` values like `"insert"` a la #4194 which could be construed as commands (these objects are not always going to be used that way.)

As secondary benefit, languages like Go have very verbose "exists in a map" syntax ([it requires multiple statements](https://golang.org/src/expvar/expvar.go#L158)), so the `new_val`/`old_val` if-else chains are even worse there, and Go has `switch` on strings, which would be very clean.

**Comment by danielmewes** [comment]
I like the `type` values @deontologician listed above as well.

**Comment by danielmewes** [comment]
In addition we should have `type: "state"` (for `includeStates`).

I'm not sure how `type: "error"` would work. Are you suggesting that we emit an event on the changefeed if someone tries to perform a write, but the write fails? I assumed #5041 was only about the `returnChanges` result, not for changefeeds. I don't see a reason for why we shouldn't be able to do the same for changefeeds though, if it's desired.

**Comment by danielmewes** [comment]
Actually I take that back:
> I don't see a reason for why we shouldn't be able to do the same for changefeeds though, if it's desired.

Usually a write fails if not enough replicas are available, or the primary isn't reachable, right? Neither case would allow us to generate corresponding changefeed events with our current architecture.
If the primary isn't available, we can't generate an event because the changefeed events are generated on the primary.
If not enough replicas are available, we only realize that after we've already emitted the change event from the primary, the way things currently work.

I was originally thinking of `unchanged` writes, that leave a document at the same value. Emitting an event for those might be possible.

**Comment by mlucy** [comment]
I think this would be OK.  Two things:
* Do we want an optarg to turn these on or off?  (I think so.)
  - Should the optarg default to on or off?  (Not sure.)
* Do we want those particular names?
  - I think the past-tense verbs like "removed" are a little weird as type names.  How would people feel about:
    * `{type: 'insert'}` or `{type: 'add'}`
    * `{type: 'delete'}` or `{type: 'remove'}`
    * `{type: 'change}`
    * `{type: initial}`
    * `{type: uninitial}`.

**Comment by deontologician** [comment]
I agree the present tense names sound better. It does seem weird to use command names like `insert` and `delete` and then call it `change` instead of `update` or `replace`.

**Comment by deontologician** [comment]
I'm less convinced there needs to be an optarg, you can already tack on `.without('type')` if you really don't like it.

**Comment by mlucy** [comment]
That's true.  Omitting the type is more efficient then adding it then removing it with `without` (quite a bit more efficient actually because `without` actually constructs a new object from scratch), but maybe the efficiency isn't worth another optarg.

**Comment by deontologician** [comment]
@danielmewes 
> I'm not sure how type: "error" would work. 

This is just me conflating `return_changes` and changefeeds. I don't think we should add an error change unless there's some other issue suggesting it.

**Comment by segphault** [comment]
I'm in favor of using an opt arg, but defaulting it to on.

Can we make it so that the strings that we use as the type values match the equivalent event names from Fusion?

**Comment by danielmewes** [comment]
> Can we make it so that the strings that we use as the type values match the equivalent event names from Fusion?

Are those documented / implemented somewhere?

**Comment by deontologician** [comment]
Right now it's `added`, `removed`, `changed` (and we don't have initial/uninitial)

**Comment by mlucy** [comment]
That might be an argument for combining `add` and `initial` into a single type and letting people who really care distinguish them by checking for the existence of an `old_val` field.  (And doing the same for `remove`/`uninitial`.)

I would guess 95% of code will want to handle them the same way, and it will avoid nasty problems like people writing code that doesn't handle `uninitial` values since they're so rare they might not show up in testing.

**Comment by encryptio** [comment]
I'd rather keep them separate. Merging `add`/`initial` and `remove`/`uninitial` makes the code to switch over types unclean again; all the languages I can think of have good support for multiple `case`s going to one codepath, but very few have the ability to switch on multiple expressions at the same time.

Also, handling `uninitial` values incorrectly is something that can still happen even if we merge it with `remove`. In either case, you'll have problems with people writing code without reading the documentation and only relying on trial and error. E.g. Python code would be buggy even in the case where they really do want to handle `uninitial` as if it were `remove` with `None`, but accessing a nonexistent key in a `dict` with `[]` raises a `KeyError`.

Overall, I think merging any event types is too subtle to be worth it. Having them separate makes people at least stop and consider that they _might_ want to do something different, and/or read the docs to clarify things.

**Comment by deontologician** [comment]
I think I agree, the point is to disambiguate them, so its better to have a simple tag that tells you unambiguously what this change represents

**Comment by encryptio** [comment]
Actually, I just checked, and I can't find an obvious mention of `uninitial` values by name or by description in either [the `.changes()` docs](https://rethinkdb.com/api/javascript/changes/) or [the "Changefeeds in RethinkDB" document](https://rethinkdb.com/docs/changefeeds/javascript/). That's probably a docs issue that needs to happen.

**Comment by mlucy** [comment]
Thinking about it a little more, we might want to leave these types off by default.  If we turn them on by default then it will encourage people to use them as their default way of interacting with changefeeds, when we want to push them to be writing code like this:

```rb
r.table('test').changes.run.each {|row|
  handle_add(row['new_val']) if row['new_val']
  handle_remove(row['old_val']) if row['old_val']
}
```

Since that's what you want your logic to be doing for 99% of cases.  Code that needs to distinguish uninitial vals from deletes is *extremely* rare, so if we leave initial/uninitial value types in as separate from added/removed value types then that would be additional motivation to leave the types off by default in order to encourage good API use.  We definitely don't want people to be writing code like this:

```rb
r.table('test').changes.run.each {|row|
  case row['type']
  when 'add'
    handle_add(row['new_val'])
  when 'initial'
    handle_add(row['new_val'])
  when 'remove'
    handle_remove(row['old_val'])
  when 'uninitial'
    handle_remove(row['old_val'])
  when 'change'
    handle_add(row['new_val'])
    handle_remove(row['old_val'])
  end
}
```

**Comment by deontologician** [comment]
The crucial use case for this, which you almost always want to do, is to distinguish between additions or removals and changes. Since it's not generally desireable to handle changes like a deletion and then an addition. You almost always want to have a separate change handler.

The following seems reasonable:

```js
switch row('type') {
  case 'initial':
  case 'add': handleAdd(row); break;
  case 'uninitial':
  case 'remove': handleRemove(row); break;
  case 'change': handleChange(row); break;
}
```
or in python (which has no switch):
```py
if row['type'] in ('initial', 'add'):
    handle_add(row)
elif row['type'] in ('uninitial', 'remove'):
    handle_remove(row)
elif row['type'] == 'change':
    handle_change(row)
```

Overall though, it seems less important how initial/uninitial are denoted.  So while I lean towards giving them their own types, I don't think we should make the tag opt-in just because uninitial is weird. We should figure out how to make it easier to deal with correctly

**Comment by mlucy** [comment]
Do you think it's more common to handle changes separately?  I mostly write testing and proof of concept code, but my code almost always handles a change the exact same as a delete followed by an insert.

**Comment by deontologician** [comment]
With leaderboards, definitely. In my case, I've written a lot of abstractions on top of changefeeds, and you just don't have the information about how they're going to be used, so the default is to split it out.

But, we aren't fundamentally changing the structure in a way that makes you choose one strategy over the other. If new_val, old_val is all you need, you can just ignore the type field. It shouldn't have many downsides

**Comment by mlucy** [comment]
Why doesn't that work for leaderboards?  (I guess this is sort of tangential to the main discussion, but I'm curious.)  Isn't it perfectly correct to remove the old value, add the new value, then rerender the leader board?

**Comment by deontologician** [comment]
If re-rendering involves re-sorting it, then it will work, but by itself you don't know where to put the new entry, you need the old one to find which item in the list to replace.

**Comment by mlucy** [comment]
That seems bad, because then you've written code that only works as long as the change doesn't modify the field you're ordering by.  If you always follow the same code path for a change as for a delete/add, though, it will work fine and handle the case where the change modifies the ordering correctly:

```rb
def order_by_field(x)
  class << x
    def <=>(other)
      x['field'] <=> other['field']
    end
  end
  x
end
lst = SortedSet.new
r.table('test').order_by(index: 'field').changes.run.each {|change|
  lst.delete(order_by_field(change['old_val'])) if change['old_val']
  lst.add(order_by_field(change['new_val'])) if change['new_val']
}
```

(If Ruby were less crazy, you'd be able to define the ordering as an argument to `SortedSet.new` instead of having to put it on the objects being passed in, but the general point is the same.)

**Comment by deontologician** [comment]
Ok, I was under the impression the ordering would always be correct by replacing the `old_val` with the `new_val`. I still think being able to go right to the changed item and modifying it is useful, vs. requiring that everyone treat `new_val` and `old_val` as completely separate events. For example, in UI's it's common to want to do animations based on a changed field.

**Comment by mlucy** [comment]
I mean, I don't think you should be *required* to treat `old_val` and `new_val` as separate events, I just think that's the right design decision most of the time and the API should push people in that direction.  (It's that "pit of success" thing @coffeemug is always talking about -- they write code that treats `new_val` and `old_val` a certain way because it's easier, then it turns out to be correct later.)

Anyway, after thinking about it for a bit I don't actually feel that strongly on it, but my current (weak) preference ordering is:
* 5 types that default to off.
* 3 types that default to on (`uninitial` is turned into `remove` and `initial` is turned into `add`).
* 5 types that default to on.
* 3 types that default to off (I think nobody wants this one).

**Comment by segphault** [comment]
In my experience writing applications, I find that you almost always want to handle changes differently from a delete/insert. Consider a scenario where you have a list of selectable items—there's potentially a lot of state on the client side that you are going to lose if you just remove the relevant item from the list and replace it. It could screw up selection, scroll position, list ordering, etc.

This is a fairly common example of what I do, taken right out of one of my demos:

```ruby
if data[:new_val] && !data[:old_val]
  # Add new item
elsif data[:new_val] && data[:old_val]
  # Update existing item
elsif !data[:new_val] && data[:old_val]
  # Remove deleted item
end
```

**Comment by danielmewes** [comment]
I think keeping the 5 different `types` is better for the reasons that were discussed.

I'm torn on whether turning this on by default is actually going to push developers towards looking at the `type` rather than just distinguishing based on whether `old_val` and `new_val` are set.

I think that either having it on or off by default would be a good solution. Everything else being equal, I'd give a slight preference for having it off by default simply because that gives us backwards compatibility.
It's also likely that users who care about handling different types will already be looking at the `changes` documentation to figure out what types events are even possible, in which case the additional `optarg` will be easy to discover.

**Comment by danielmewes** [comment]
For the names, my vote is for
`{type: 'add'}`
`{type: 'remove'}`
`{type: 'change'}`
`{type: 'initial'}`
`{type: 'uninitial'}`

We should probably just change the Fusion names to align with this.

**Comment by danielmewes** [comment]
Also for the record ( @encryptio ): The docs issue for explaining uninitial values is https://github.com/rethinkdb/docs/issues/966

**Comment by deontologician** [comment]
My preference order is:

5 types that default to on
3 types that default to on
5 types that default to off

I don't default on presents much of a backwards compatibility issue since it's a new key. It's unlikely anyone is making assertions that keys aren't present. Also, I think it's likely to be the other way around with reading documentation, where you see the types in the data coming back and now you have something to google for

**Comment by mlucy** [comment]
Looks like there's no real consensus with just us 3 then.  Anyone else have an ordering of which options they'd prefer?

**Comment by coffeemug** [comment]
TL;DR: I think we should drop/deprecate `includes_states`, add an `output` optarg, and send 5 types.

I really like what this issue is trying to accomplish, but I'm concerned including a `type` field into existing feeds might make them *more* confusing (especially if it's controlled by an optarg).

We already have an `include_states` optarg that injects a `state` field. Now we'll have an `includes_types` optarg that will inject a `type` field, and the information in the `type` field will partially intersect with the information that the `state` field already provides. I think that would leave users with too many unaswered questions (e.g. "what's the difference between `type` and `state`? when do I use each? why do both exist?")

If we make types on by default, then there is even more confusion (e.g. "oh, so `types` are on by default, `states` are off by default, and the difference between them is...?")

I think a better option would be to replace the `includes_states` optarg with an `output` optarg that takes on multiple values. For now, these values could be `verbose` (default), and `compact`.

- `verbose` would be equivalent to what is proposed here (i.e. will include the types *instead* of states) and would be the default.
- `compact` would be equivalent to what currently is accomplished via `includes_states=True`.
- We might also add something like `extra_compact` (but with a better name) that would be equivalent to `include_states=False`, though I don't think this format is very useful since it loses information.

Eventually, we'll implement #3719 and #4707 and `output` will be extended to support more formats (e.g. `all/list` and `diff`). This seems much cleaner than the existing proposal of bolting on `includes_types`.

As far as merging `add/initial` and `remove/uninitial`, it seems that everyone who commented in this thread thinks it's a bad idea. I actually do think that 95% of the time users will want to follow the progression of:

- Just define add/remove handlers and have the system translate other events to these
- Define the change handler for specialized behavior
- Define the initial/uninitial handlers for specialized behavior

I think this sort of behavior is super-valuable, but should be up to the drivers. I wouldn't mess with the types in the database, so I prefer not merging events into a single type on the server level.

**Comment by mlucy** [comment]
@coffeemug -- I don't think I understand what you're proposing.  You can't include types "instead" of states because they contain different information.  For example, let's say you write `r.table('test').changes(include_initial: true)` and you have types but no states, and no writes occur -- how do you know when you've read the last document?  Knowing that the type of every event is `initial` doesn't tell you when you're done, but a `{state: 'ready'}` document would.

In general types and states convey completely different things: the type tells you something about a particular change, while the state tells you something about the stream of changes as a whole.

**Comment by danielmewes** [comment]
I'm on the same page with @mlucy here. I think `state` and `type` are very different things. In fact you'd never get a document with both fields according to the current proposal (though see below).

I can see the appeal of grouping multiple changefeed options into easier to understand "presets"/verbosity levels, but I don't think those should be the only way to configure the options. I can see a lot of use cases where users won't be interested in the `type` field, but would still want `state` documents for example.
At the same time, I don't think we're having enough options yet to warrant having the verbosity setting *in addition* to the individual arguments. I think that would only make the API more confusing.


Unrelated to this, I think we should also have a `type: "state"` value that we put into state documents when `include_states` is set. I forgot about those in the list of `type` values above.

**Comment by coffeemug** [comment]
Grrr, that's embarrassing. It looks like everything I said last night was wrong.

Ok, in that case, I'd vote for 5 types that default to off to be congruent with `include_states`.

**Comment by mlucy** [comment]
It sounds like we have a slight majority in favor of 5 types that default to off.  (That also seems like the safer option, because it will break literally 0 existing code even if somebody is counting the keys in the change documents for some reason, maybe to distinguish inserts from initial vals.)

**Comment by mlucy** [comment]
It would also be pretty easy to change the default to `on` if it turns out literally everybody is always turning them on.  (Whereas if the default is `on` it would be break a *lot* of code to change the default to `off`.)

**Comment by deontologician** [comment]
It won't be terrible to add an option, and it's certainly better than not
getting the tags at all. I'm fine with 5/off if that's the consensus
otherwise

**Comment by encryptio** [comment]
A bunch of little things (incremental API improvements, implying use of `old_val`/`new_val` null-ness is better if useful, and consistency with the other options being default-off optargs) now add up to me now preferring 5/off for this. As @mlucy said, moving to 5/on later (my second preference) is an easy, compatible change to make.

**Comment by danielmewes** [comment]
I think there's a majority for off by default.
Marking as settled for 6/off (the 5 ones listed above in present tense + `"state"`).

**Comment by nighelles** [comment]
This is in review #3525.
Implemented in nighelles/5188

**Comment by nighelles** [comment]
This is in next.
